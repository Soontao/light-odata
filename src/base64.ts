/* eslint-disable max-len */
/* eslint-disable prefer-const */
// Array of bytes to Base64 string decoding
function b64ToUint6(nChr) {
  return nChr > 64 && nChr < 91 ?
    nChr - 65
    : nChr > 96 && nChr < 123 ?
      nChr - 71
      : nChr > 47 && nChr < 58 ?
        nChr + 4
        : nChr === 43 ?
          62
          : nChr === 47 ?
            63
            :
            0;
}

function base64DecToArr(sBase64: string, nBlocksSize?: number) {
  const sB64Enc = sBase64.replace(/[^A-Za-z0-9+/]/g, "");
  const nInLen = sB64Enc.length;
  const nOutLen = nBlocksSize ? Math.ceil((nInLen * 3 + 1 >> 2) / nBlocksSize) * nBlocksSize : nInLen * 3 + 1 >> 2;
  const taBytes = new Uint8Array(nOutLen);

  let nMod3;
  let nMod4;
  let nUint24 = 0;
  let nOutIdx = 0;
  for (let nInIdx = 0; nInIdx < nInLen; nInIdx++) {
    nMod4 = nInIdx & 3;
    nUint24 |= b64ToUint6(sB64Enc.charCodeAt(nInIdx)) << 6 * (3 - nMod4);
    if (nMod4 === 3 || nInLen - nInIdx === 1) {
      nMod3 = 0;
      while (nMod3 < 3 && nOutIdx < nOutLen) {
        taBytes[nOutIdx] = nUint24 >>> (16 >>> nMod3 & 24) & 255;
        nMod3++;
        nOutIdx++;
      }
      nUint24 = 0;

    }
  }

  return taBytes;
}

/* Base64 string to array encoding */
function uint6ToB64(nUint6: number) {
  return nUint6 < 26 ?
    nUint6 + 65
    : nUint6 < 52 ?
      nUint6 + 71
      : nUint6 < 62 ?
        nUint6 - 4
        : nUint6 === 62 ?
          43
          : nUint6 === 63 ?
            47
            :
            65;
}

function base64EncArr(aBytes) {
  let nMod3 = 2;
  let sB64Enc = "";

  const nLen = aBytes.length;
  let nUint24 = 0;
  for (let nIdx = 0; nIdx < nLen; nIdx++) {
    nMod3 = nIdx % 3;
    if (nIdx > 0 && (nIdx * 4 / 3) % 76 === 0) {
      sB64Enc += "\r\n";
    }

    nUint24 |= aBytes[nIdx] << (16 >>> nMod3 & 24);
    if (nMod3 === 2 || aBytes.length - nIdx === 1) {
      sB64Enc += String.fromCodePoint(uint6ToB64(nUint24 >>> 18 & 63), uint6ToB64(nUint24 >>> 12 & 63), uint6ToB64(nUint24 >>> 6 & 63), uint6ToB64(nUint24 & 63));
      nUint24 = 0;
    }
  }
  return sB64Enc.substring(0, sB64Enc.length - 2 + nMod3) + (nMod3 === 2 ? "" : nMod3 === 1 ? "=" : "==");
}

/* UTF-8 array to JS string and vice versa */

function utf8ArrToStr(aBytes) {
  let sView = "";
  let nPart;
  const nLen = aBytes.length;
  for (let nIdx = 0; nIdx < nLen; nIdx++) {
    nPart = aBytes[nIdx];
    sView += String.fromCodePoint(
      nPart > 251 && nPart < 254 && nIdx + 5 < nLen ? /* six bytes */
        /* (nPart - 252 << 30) may be not so safe in ECMAScript! So…: */
        (nPart - 252) * 1073741824 + (aBytes[++nIdx] - 128 << 24) + (aBytes[++nIdx] - 128 << 18) + (aBytes[++nIdx] - 128 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
        : nPart > 247 && nPart < 252 && nIdx + 4 < nLen ? /* five bytes */
          (nPart - 248 << 24) + (aBytes[++nIdx] - 128 << 18) + (aBytes[++nIdx] - 128 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
          : nPart > 239 && nPart < 248 && nIdx + 3 < nLen ? /* four bytes */
            (nPart - 240 << 18) + (aBytes[++nIdx] - 128 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
            : nPart > 223 && nPart < 240 && nIdx + 2 < nLen ? /* three bytes */
              (nPart - 224 << 12) + (aBytes[++nIdx] - 128 << 6) + aBytes[++nIdx] - 128
              : nPart > 191 && nPart < 224 && nIdx + 1 < nLen ? /* two bytes */
                (nPart - 192 << 6) + aBytes[++nIdx] - 128
                : /* nPart < 127 ? */ /* one byte */
                nPart
    );
  }
  return sView;
}

function strToUTF8Arr(sDOMStr: string): Uint8Array {
  let aBytes;
  let nChr;
  const nStrLen = sDOMStr.length;
  let nArrLen = 0;

  /* mapping… */
  for (let nMapIdx = 0; nMapIdx < nStrLen; nMapIdx++) {
    nChr = sDOMStr.codePointAt(nMapIdx);

    if (nChr > 65536) {
      nMapIdx++;
    }

    nArrLen += nChr < 0x80 ? 1 : nChr < 0x800 ? 2 : nChr < 0x10000 ? 3 : nChr < 0x200000 ? 4 : nChr < 0x4000000 ? 5 : 6;
  }

  aBytes = new Uint8Array(nArrLen);

  /* transcription… */
  let nIdx = 0;
  let nChrIdx = 0;
  while (nIdx < nArrLen) {
    nChr = sDOMStr.codePointAt(nChrIdx);
    if (nChr < 128) {
      /* one byte */
      aBytes[nIdx++] = nChr;
    } else if (nChr < 0x800) {
      /* two bytes */
      aBytes[nIdx++] = 192 + (nChr >>> 6);
      aBytes[nIdx++] = 128 + (nChr & 63);
    } else if (nChr < 0x10000) {
      /* three bytes */
      aBytes[nIdx++] = 224 + (nChr >>> 12);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
    } else if (nChr < 0x200000) {
      /* four bytes */
      aBytes[nIdx++] = 240 + (nChr >>> 18);
      aBytes[nIdx++] = 128 + (nChr >>> 12 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
      nChrIdx++;
    } else if (nChr < 0x4000000) {
      /* five bytes */
      aBytes[nIdx++] = 248 + (nChr >>> 24);
      aBytes[nIdx++] = 128 + (nChr >>> 18 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 12 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
      nChrIdx++;
    } else /* if (nChr <= 0x7fffffff) */ {
      /* six bytes */
      aBytes[nIdx++] = 252 + (nChr >>> 30);
      aBytes[nIdx++] = 128 + (nChr >>> 24 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 18 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 12 & 63);
      aBytes[nIdx++] = 128 + (nChr >>> 6 & 63);
      aBytes[nIdx++] = 128 + (nChr & 63);
      nChrIdx++;
    }
    nChrIdx++;
  }

  return aBytes;
}

const encodeJs = (input: string) => {
  return base64EncArr(strToUTF8Arr(input));
};

const decodeJs = (input: string) => {
  return utf8ArrToStr(base64DecToArr(input));
};

/**
 * encode string to base64
 *
 * @param input
 */
const encode = function (input = ""): string {
  if (globalThis.Buffer !== undefined) {
    return globalThis.Buffer.from(input, "utf-8").toString("base64");
  }
  return base64EncArr(strToUTF8Arr(input));
};

/**
 * decode base64 to string
 *
 * @param input base64 string
 */
const decode = function (input: string): string {
  if (globalThis.Buffer !== undefined) {
    return globalThis.Buffer.from(input, "base64").toString("utf-8");
  }
  return utf8ArrToStr(base64DecToArr(input));
};

export { encode, decode, encodeJs, decodeJs };
