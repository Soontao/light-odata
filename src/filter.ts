/* eslint-disable max-len */
import { join } from "@newdash/newdash/join";
import { FrameworkError, ValidationError } from "./errors";
import { SystemQueryOptions } from "./params";
import { ODataDateTime, ODataDateTimeOffset, ODataValueObject } from "./types";


export type FilterValue = number | string | ODataValueObject | null | boolean | Symbol | object

export enum ExprOperator {
  eq = "eq",
  ne = "ne",
  gt = "gt",
  lt = "lt",
  ge = "ge",
  le = "le",
}

type FieldExpr = {
  op: ExprOperator;
  value: FilterValue;
}

type FieldExprMappings = {
  [key: string]: FieldExpr[]
}

/**
 * odata property filter expression
 * @private
 * @internal
 */
class ODataPropertyExpr<T extends ODataFilter> {

  constructor(filter: T, fieldName: string, mapping: FieldExprMappings) {
    this._exprMappings = mapping;
    this._fieldName = fieldName;
    this._filter = filter;
    // initilize
    if (this._getFieldExprs() == undefined) {
      this._exprMappings[this._fieldName] = [];
    }
  }

  private _filter: T;

  private _fieldName: string;

  private _exprMappings: FieldExprMappings;

  private _getFieldExprs() {
    return this._exprMappings[this._fieldName];
  }

  private _addExpr(op: ExprOperator, value: FilterValue) {

    switch (typeof value) {
      case "number": case "boolean":
        this._getFieldExprs().push({ op, value: `${value}` });
        break;
      case "string":
        if (value.startsWith("'") || value.startsWith("datetime")) {
          this._getFieldExprs().push({ op, value });
        } else {
          this._getFieldExprs().push({ op, value: `'${value}'` });
        }
        break;
      case "object":
        if (value instanceof ODataValueObject) {
          this._getFieldExprs().push({ op, value: value.toString() });
        } else if (value === null) {
          this._getFieldExprs().push({ op, value: null });
        } else {
          throw new FrameworkError(`Not support object ${value?.["constructor"]?.["name"] || typeof value} in odata filter eq/ne/gt/ge/ne/nt ...`);
        }
        break;
      case "undefined":
        throw new ValidationError(`You must set value in odata filter eq/ne/gt/ge/ne/nt ...`);
      default:
        throw new FrameworkError(`Not support typeof ${typeof value}: ${value} in odata filter eq/ne/gt/ge/ne/nt ...`);
    }

  }

  /**
   * equal
   * @param value
   */
  eq(value: FilterValue): T {
    this._addExpr(ExprOperator.eq, value);
    return this._filter;
  }

  /**
   * not equal
   * @param value
   */
  ne(value: FilterValue): T {
    this._addExpr(ExprOperator.ne, value);
    return this._filter;
  }

  /**
   * @deprecated use `filter.eq("value")` instead
   * @param value
   * @returns
   */
  eqString(value: string): T {
    this._addExpr(ExprOperator.eq, `'${value}'`);
    return this._filter;
  }

  /**
   * @deprecated use `filter.ne("value")` instead
   * @param value
   * @returns
   */
  neString(value: string): T {
    this._addExpr(ExprOperator.ne, `'${value}'`);
    return this._filter;
  }

  /**
   * greater or equal
   * @param value
   */
  ge(value: FilterValue): T {
    this._addExpr(ExprOperator.ge, value);
    return this._filter;
  }

  /**
   * greater than
   * @param value
   */
  gt(value: FilterValue): T {
    this._addExpr(ExprOperator.gt, value);
    return this._filter;
  }

  /**
   * less or equal
   * @param value
   */
  le(value: FilterValue): T {
    this._addExpr(ExprOperator.le, value);
    return this._filter;
  }

  /**
   * less than
   * @param value
   */
  lt(value: FilterValue): T {
    this._addExpr(ExprOperator.lt, value);
    return this._filter;
  }

  /**
   * match any value in an array
   *
   * @param values
   */
  in(values: FilterValue[] = []): T {
    if (values.length > 0) {
      values.forEach((value) => {
        this.eq(value);
      });
    }
    return this._filter;
  }

  /**
   * filter by value range
   *
   * @param low
   * @param max
   * @param includeBoundary
   */
  between(low: FilterValue, max: FilterValue, includeBoundary = true): T {
    if (low == undefined || max == undefined) {
      throw new ValidationError("You must give out the start and end value");
    }
    if (includeBoundary) {
      this.ge(low);
      this.le(max);
    } else {
      this.gt(low);
      this.lt(max);
    }
    return this._filter;
  }

  /**
   *
   * @deprecated please use filter.between(ODataDateTime.from(date), ODataDateTime.from(date))
   * @param start
   * @param end
   * @param includeBoundary
   * @returns
   */
  betweenDateTime(start?: Date, end?: Date, includeBoundary = true): T {
    if (start == undefined && end == undefined) {
      throw new ValidationError("You must give out the start or end date");
    }
    if (start instanceof Date) {
      if (includeBoundary) {
        this.ge(ODataDateTime.from(start));
      } else {
        this.gt(ODataDateTime.from(start));
      }
    }
    if (end instanceof Date) {
      if (includeBoundary) {
        this.le(ODataDateTime.from(end));
      } else {
        this.lt(ODataDateTime.from(end));
      }
    }
    return this._filter;
  }

  /**
   *  @deprecated please use filter.between(ODataDateTimeOffset.from(date), ODataDateTimeOffset.from(date))
   * @param start
   * @param end
   * @param includeBoundary
   * @returns
   */
  betweenDateTimeOffset(start?: Date, end?: Date, includeBoundary = true): T {
    if (start == undefined && end == undefined) {
      throw new ValidationError("You must give out the start or end date");
    }
    if (start instanceof Date) {
      if (includeBoundary) {
        this.ge(ODataDateTimeOffset.from(start));
      } else {
        this.gt(ODataDateTimeOffset.from(start));
      }
    }
    if (end instanceof Date) {
      if (includeBoundary) {
        this.le(ODataDateTimeOffset.from(end));
      } else {
        this.lt(ODataDateTimeOffset.from(end));
      }
    }
    return this._filter;
  }

}


/**
 * OData filter builder
 *
 */
export class ODataFilter<T = any> {

  static newBuilder<T = any>() {
    return new ODataFilter<T>();
  }

  /**
   * construct a new filter
   */
  static newFilter<T = any>() {
    return new ODataFilter<T>();
  }

  private _fieldExprMappings: FieldExprMappings = {};

  /**
   * getExprMapping
   *
   * @internal
   * @private
   */
  private getExprMapping(): FieldExprMappings {
    return this._fieldExprMappings;
  }

  /**
   * filter property
   * @see property
   * @param name field name or function expression
   */
  field(name: keyof T): ODataPropertyExpr<this> {
    return this.property(name);
  }

  /**
   * filter property
   *
   * @param name property name
   */
  property(name: keyof T): ODataPropertyExpr<this> {
    // @ts-ignore
    return new ODataPropertyExpr(this, name, this.getExprMapping());
  }

  /**
   * The value of a field matches any value in the list.
   *
   * @deprecated please use filter.field().in()
   * @param name
   * @param values
   */
  fieldIn(name: keyof T, values: string[]): this {
    // @ts-ignore
    return this.fieldValueMatchArray(name, values);
  }

  /**
   * The value of a field matches any value in the list.
   *
   * @deprecated please use filter.field().in()
   * @param name
   * @param values
   */
  fieldValueMatchArray(name: keyof T, values: string[] = []): this {
    if (values) {
      values.forEach((value) => {
        this.field(name).eqString(value);
      });
    }
    return this;
  }

  /**
   * DEPRECATED
   *
   * please use betweenDateTime/betweenDateTimeOffset
   *
   * @deprecated
   * @param name
   * @param start
   * @param end
   */
  inPeriod(name: keyof T, start: Date, end: Date): this {
    return this.betweenDateTime(name, start, end);
  }

  /**
   * @deprecated
   * @param name
   * @param start
   * @param end
   */
  betweenDateTime(name: keyof T, start: Date, end: Date): this {
    if (start && end) {
      return this.gtDateTime(name, start).ltDateTime(name, end);
    }
    throw new ValidationError("You must give out the start and end date");

  }

  /**
   * @deprecated
   * @param name
   * @param start
   * @param end
   */
  betweenDateTimeOffset(name: keyof T, start: Date, end: Date): ODataFilter {
    if (start && end) {
      return this.gtDateTimeOffset(name, start).ltDateTimeOffset(name, end);
    }
    throw new ValidationError("You must give out the start and end date");
  }

  /**
   * @deprecated
   * @param name
   * @param date
   */
  gtDateTime(name: keyof T, date: Date): this {
    return this.field(name).gt(ODataDateTime.from(date));
  }

  /**
   * @deprecated
   * @param name
   * @param date
   */
  gtDateTimeOffset(name: keyof T, date: Date): this {
    return this.field(name).gt(ODataDateTimeOffset.from(date));
  }

  /**
   * @deprecated
   * @param name
   * @param date
   */
  ltDateTime(name: keyof T, date: Date): this {
    return this.field(name).lt(ODataDateTime.from(date));
  }

  /**
   * @deprecated
   * @param name
   * @param date
   */
  ltDateTimeOffset(name: keyof T, date: Date): this {
    return this.field(name).lt(ODataDateTimeOffset.from(date));
  }

  /**
   * @deprecated c4codata will auto group exprs
   * @param filter
   */
  group(filter: ODataFilter): this {
    this._fieldExprMappings = Object.assign(this._fieldExprMappings, filter.getExprMapping());
    return this;
  }

  toString(): string {
    return this.build();
  }

  protected _buildFieldExprString(field: string): string {
    const exprs = this.getExprMapping()[field];
    if (exprs.length > 0) {
      if (exprs.filter((expr) => expr.op == ExprOperator.eq).length == 0) {
        return `(${join(
          exprs.map(({ op, value }) => `${field} ${op} ${value}`),
          " and "
        )})`;
      }
      return `(${join(
        exprs.map(({ op, value }) => `${field} ${op} ${value}`),
        " or "
      )})`;

    }
    return "";


  }

  /**
   * build filter to string
   */
  build(): string {
    let _rt = "";
    _rt = join(
      // join all fields exprs string
      Object.entries(this.getExprMapping()).map(
        ([fieldName, exprs]) => {
          switch (exprs.length) {
            // if one field expr mapping array is empty
            case 0:
              return "";
            // only have one expr
            case 1:
              const { op, value } = exprs[0];
              return `${fieldName} ${op} ${value}`;
            default:
              // multi exprs
              return this._buildFieldExprString(fieldName);
          }

        }
      ),
      " and "
    );
    return _rt;
  }

}

export class ParamBoundedFilter<T = any> extends ODataFilter<T> {

  #params: SystemQueryOptions;

  constructor(params: SystemQueryOptions) {
    super();
    this.#params = params;
  }

  /**
   * the filter is ended, return the param object
   *
   * @returns the param
   */
  public filterEnd(): SystemQueryOptions {
    return this.#params.filter(this);
  }

}

export function filter() {
  return ODataFilter.newFilter();
}
