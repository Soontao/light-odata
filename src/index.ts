import { OData } from "./odata";

export * from "./batch";
export * from "./filter";
export * from "./params";
export * from "./types";
export { EdmV4 } from "./types_v4";
export * from "./util";
export * from "./request";
export { transformation } from "./tranformation";
export { OData };

export default OData;
