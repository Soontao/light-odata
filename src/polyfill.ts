global.fetch = require("fetch-cookie")(require("node-fetch"));
