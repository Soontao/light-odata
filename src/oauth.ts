import { Mutex } from "@newdash/newdash/functional/Mutex";
import { encode } from "./base64";
import { AuthenticationError } from "./errors";
import { SearchParams } from "./util";

const S_AUTHORIZATION = "Authorization";
const S_BEARER = "Bearer";
const S_CLIENT_CREDENTIALS = "client_credentials";
const S_CT_URL_FORM = "application/x-www-form-urlencoded";

export type TokenRetrieveType = "header" | "form";

/**
 * Client Credential OAuth Client
 *
 * Will automatically process the token refresh logic
 *
 * refresh policy: half expire-in
 *
 * @example
 * ```js
 * // please remember share client instance for a same OAuth server
 * const client = new ClientCredentialsOAuthClient(
 *  "http://example/oauth/token",
 *  "client_id",
 *  "client_secret"
 * )
 * const response = await client.fetch("http://protected.by.token/resource", {
 *  method: "GET",
 * })
 * ```
 *
 *
 */
class ClientCredentialsOAuthClient {

  private readonly tokenUrl: string;

  private readonly clientId: string;

  private readonly clientSecret: string;

  private readonly scope: string;

  private readonly mut = new Mutex();

  private token: string;

  private expiresAt = 0;

  private tokenRetrieveType: TokenRetrieveType = "header";


  /**
   * a very simple oauth client credential client
   *
   * @param tokenUrl oauth token url
   * @param clientId oauth client id
   * @param clientSecret oauth client secret
   * @param retrieveType the clientId and clientSecret is put into header or form body
   * @param scope oauth scope
   */
  constructor(
    tokenUrl: string,
    clientId: string,
    clientSecret: string,
    retrieveType: TokenRetrieveType = "header",
    scope?: string,
  ) {

    this.tokenUrl = tokenUrl;
    this.clientId = clientId;
    this.clientSecret = clientSecret;
    this.tokenRetrieveType = retrieveType;
    this.scope = scope;
  }

  /**
   * fetch token directly
   * @private
   * @returns {Promise<any>}
   */
  private async fetchOAuthResponse(): Promise<any> {
    const params = new SearchParams();

    params.append("grant_type", S_CLIENT_CREDENTIALS);

    if (typeof this.scope === "string") { params.append("scope", this.scope); }

    let response = undefined;

    if (this.tokenRetrieveType === "form") {

      params.append("client_id", this.clientId);
      params.append("client_secret", this.clientSecret);

      response = await fetch(this.tokenUrl, {
        method: "POST",
        body: params.toString(),
        headers: { "Content-Type": S_CT_URL_FORM }
      });

    }
    else {

      // standard way described in rfc6749
      // https://datatracker.ietf.org/doc/html/rfc6749#section-4.4.2
      response = await fetch(this.tokenUrl, {
        method: "POST",
        body: params.toString(),
        headers: {
          [S_AUTHORIZATION]: "Basic " + encode(`${this.clientId}:${this.clientSecret}`),
          "Content-Type": S_CT_URL_FORM
        }
      });

    }


    if (response.status >= 400) {

      if (response.headers.get("Content-Type")?.includes?.("application/json")) {
        const responseBody = await response.json();
        if ("error" in responseBody) {
          throw new AuthenticationError(responseBody.error);
        }
        else {
          throw new AuthenticationError(JSON.stringify(responseBody));
        }
      }
      else {
        throw new Error(await response.text());
      }
    }

    return response.json();
  }

  /**
   * get header pair
   */
  public async getHeader() {
    const token = await this.getToken();
    return {
      [S_AUTHORIZATION]: `${S_BEARER} ${token}`
    };
  }

  /**
   * get current valid oauth token
   *
   * @returns {Promise<string>}
   */
  public async getToken(): Promise<string> {

    // lock avoid multi fetch token requests in same time
    return this.mut.use(async () => {

      try {

        const currentTime = new Date().getTime();

        // retrive new token
        if (currentTime >= this.expiresAt) {
          const { access_token: accessToken, expires_in: expiresIn } = await this.fetchOAuthResponse();
          this.token = accessToken;
          // half expire
          this.expiresAt = currentTime + ((expiresIn / 2) * 1000);
        }

      } catch (error) {

        console?.error?.(`oauth client retrieve token failed: ${error?.message ?? error}`,);

        // rethrow error
        throw error;

      }

      return this.token;

    });


  }

  /**
   * send requests with oauth token
   *
   */
  public async fetch(...args: Parameters<typeof fetch>) {
    // refresh token if required
    const token = await this.getToken();

    if (token === undefined) {
      throw new TypeError("un-expected token is undefined");
    }

    const url = args[0];
    let config = args[1];

    if (config === undefined) {
      config = {};
    }
    // append jwt token
    config.headers = {
      ...(config.headers || {}),
      [S_AUTHORIZATION]: `${S_BEARER} ${token}`
    };

    return fetch(url, config);
  }

}

export { ClientCredentialsOAuthClient };
