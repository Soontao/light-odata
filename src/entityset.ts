// @ts-nocheck
import { ODataServerError } from "./errors";
import { ODataFilter } from "./filter";
import { systemOptions, SystemQueryOptions } from "./params";
import { OData } from "./odata";
import type { DeepPartial, ODataKeyPredicate, PlainODataResponse } from "./types";

/**
 * EntitySet
 */
export class EntitySet<T = any> {

  private _entitySet: string;

  private _client: OData;

  constructor(collection: string, client: OData) {
    this._entitySet = collection;
    this._client = client;
  }

  private _checkError(res: any): void {
    if (res.error) {
      switch (this._client.getVersion()) {
        case "v2":
          throw new ODataServerError(res.error?.message?.value);
        case "v4":
          throw new ODataServerError(res.error?.message);
        default:
          break;
      }
    }
  }

  private _getResultSingle(res: PlainODataResponse) {
    switch (this._client.getVersion()) {
      case "v2":
        // @ts-ignore
        return res.d;
      case "v4":
        // @ts-ignore
        return res;
      default:
        break;
    }
  }

  private _getResults(res: PlainODataResponse) {
    switch (this._client.getVersion()) {
      case "v2":
        // @ts-ignore
        return res.d?.results;
      case "v4":
        // @ts-ignore
        return res?.value;
      default:
        break;
    }
  }

  public newFilter(): ODataFilter<T> {
    return this._client.newFilter();
  }

  /**
   * @deprecated
   * @returns
   */
  public newParam(): SystemQueryOptions<T> {
    return this._client.newParam();
  }

  public newOptions() {
    return this._client.newOptions();
  }

  async retrieve(id: ODataKeyPredicate, systemOptions?: SystemQueryOptions): Promise<T> {
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "GET",
      id,
      params: systemOptions
    });
    this._checkError(res);
    return this._getResultSingle(res);
  }


  async find(base: DeepPartial<T>): Promise<T[]> {
    const filter = OData.newFilter();

    Object.entries(base).forEach(([key, value]) => {
      if (typeof value == "string") {
        filter.field(key).eqString(value);
      } else {
        filter.field(key).eq(value);
      }
    });

    return this.query(OData.newParam().filter(filter));
  }

  async query(filter: ODataFilter): Promise<T[]>;

  async query(systemOptions: SystemQueryOptions): Promise<T[]>;

  async query(): Promise<T[]>;

  async query(param: any): Promise<any> {
    if (param instanceof ODataFilter) {
      param = systemOptions().filter(param);
    }
    if (param == undefined) {
      param = OData.newParam();
    }
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "GET",
      params: param
    });
    this._checkError(res);
    return this._getResults(res);
  }

  async count(filter?: ODataFilter): Promise<number>;

  async count(filter?: Partial<T>): Promise<number>;

  async count(filter?: any) {

    const params = OData.newParam().top(1);

    // set count flag
    switch (this._client.getVersion()) {
      case "v4":
        params.count(true);
        break;
      case "v2":
        params.inlinecount(true);
        break;
    }

    if (typeof filter === "object") {

      if (filter instanceof ODataFilter) {
        params.filter(filter);
      } else {
        const newFilter = this.newFilter();
        for (const [key, value] of Object.entries(filter)) {
          newFilter.field(key).eq(value);
        }
        params.filter(newFilter);
      }

    }

    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "GET",
      params
    });

    this._checkError(res);
    switch (this._client.getVersion()) {
      case "v2":
        return parseInt(res?.d?.__count);
      case "v4":
        return res["@odata.count"];
    }
  }

  async create(body: DeepPartial<T>): Promise<T>;

  async create(body: any): Promise<T>;

  async create(body: DeepPartial<T>): Promise<T> {
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "POST",
      entity: body
    });
    this._checkError(res);
    return this._getResultSingle(res);

  }

  /**
   * update instance by `UPDATE` method
   *
   * @param id
   * @param body
   */
  async update(id: ODataKeyPredicate, body: DeepPartial<T>): Promise<void> {
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "PATCH",
      id,
      entity: body
    });
    this._checkError(res);
  }

  /**
   * delete instance by object keys
   *
   * @param id
   */
  async delete(id: ODataKeyPredicate): Promise<void> {
    const res = await this._client.newRequest<T>({
      collection: this._entitySet,
      method: "DELETE",
      id
    });
    this._checkError(res);
  }

  /**
   * perform bounded action
   *
   * @param actionName the action name, remember add namespace for it
   * @param parameters
   */
  async action(actionName: string, id: ODataKeyPredicate, parameters?: any): any {
    const responseBody = await this._client.newRequest({
      collection: this._entitySet,
      method: "POST",
      id,
      parameters,
      actionName
    });
    this._checkError(responseBody);
    return responseBody;
  }

  /**
   * perform bounded function
   *
   * @param functionName
   * @param id
   */
  async function(
    functionName: string,
    id: ODataKeyPredicate,
    parameters?: any,
    systemOptions?: SystemQueryOptions
  ): any {
    const responseBody = await this._client.newRequest({
      collection: this._entitySet,
      method: "GET",
      id,
      params: systemOptions,
      parameters,
      functionName
    });
    this._checkError(responseBody);
    return responseBody;
  }



}
