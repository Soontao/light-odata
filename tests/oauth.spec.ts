// @ts-nocheck
import { Headers } from "node-fetch";
import { ClientCredentialsOAuthClient } from "../src/oauth";

const fetch = global.fetch = jest.fn()

describe('OAuth2 Client Credential Test Suite', () => {

  const TEST_OAUTH_TOKEN_URL = 'https://a.b.c.d/oauth2/token'
  const TEST_OAUTH_CLIENT_ID = 'client_id'
  const TEST_OAUTH_CLIENT_SECRET = 'client_secret'

  beforeEach(() => {
    fetch.mockReset()
    fetch.mockResolvedValue({ status: 200, json: jest.fn().mockResolvedValue({ access_token: 'token1', expires_in: 120 }) })

  })

  it('should support retrieve token', async () => {

    const client = new ClientCredentialsOAuthClient(
      TEST_OAUTH_TOKEN_URL,
      TEST_OAUTH_CLIENT_ID,
      TEST_OAUTH_CLIENT_SECRET
    )

    const fetchFunctionSpy = jest.spyOn(client, "fetchOAuthResponse")

    const token1 = await client.getToken()
    expect(fetchFunctionSpy).toBeCalled()
    expect(token1).not.toBeUndefined()
    expect(typeof token1).toBe("string")

    expect(fetch.mock.lastCall).toMatchSnapshot()

    const token2 = await client.getToken()
    expect(fetchFunctionSpy).toBeCalledTimes(1)
    expect(token2).toBe(token1)

    await client.fetch("http://whatevery.acc.com/resource")
  });

  it('should support retrieve token by form', async () => {

    const client = new ClientCredentialsOAuthClient(
      TEST_OAUTH_TOKEN_URL,
      TEST_OAUTH_CLIENT_ID,
      TEST_OAUTH_CLIENT_SECRET,
      "form"
    )

    const fetchFunctionSpy = jest.spyOn(client, "fetchOAuthResponse")

    const token1 = await client.getToken()
    expect(fetchFunctionSpy).toBeCalled()
    expect(token1).not.toBeUndefined()
    expect(typeof token1).toBe("string")

    expect(fetch.mock.lastCall).toMatchSnapshot()

    const token2 = await client.getToken()
    expect(fetchFunctionSpy).toBeCalledTimes(1)
    expect(token2).toBe(token1)

  });

  it('should reject error when authenticated failed', async () => {
    const client = new ClientCredentialsOAuthClient(
      TEST_OAUTH_TOKEN_URL,
      TEST_OAUTH_CLIENT_ID,
      "wrong"
    )

    fetch.mockResolvedValueOnce({
      status: 401,
      headers: new Headers({ "Content-Type": 'application/json' }),
      json: jest.fn().mockResolvedValue({ error: "unauthorized" })
    })

    await expect(() => client.getToken())
      .rejects
      .toThrow("unauthorized")
    expect(fetch.mock.lastCall).toMatchSnapshot()

  });

});
