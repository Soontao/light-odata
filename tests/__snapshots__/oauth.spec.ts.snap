// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OAuth2 Client Credential Test Suite should reject error when authenticated failed 1`] = `
[
  "https://a.b.c.d/oauth2/token",
  {
    "body": "grant_type=client_credentials",
    "headers": {
      "Authorization": "Basic Y2xpZW50X2lkOndyb25n",
      "Content-Type": "application/x-www-form-urlencoded",
    },
    "method": "POST",
  },
]
`;

exports[`OAuth2 Client Credential Test Suite should support retrieve token 1`] = `
[
  "https://a.b.c.d/oauth2/token",
  {
    "body": "grant_type=client_credentials",
    "headers": {
      "Authorization": "Basic Y2xpZW50X2lkOmNsaWVudF9zZWNyZXQ=",
      "Content-Type": "application/x-www-form-urlencoded",
    },
    "method": "POST",
  },
]
`;

exports[`OAuth2 Client Credential Test Suite should support retrieve token by form 1`] = `
[
  "https://a.b.c.d/oauth2/token",
  {
    "body": "grant_type=client_credentials&client_id=client_id&client_secret=client_secret",
    "headers": {
      "Content-Type": "application/x-www-form-urlencoded",
    },
    "method": "POST",
  },
]
`;
